File: glide/gestures/velocity_controller.py
"""Simplified velocity-based scroll controller."""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum

from glide.gestures.velocity_tracker import Vec2D


class GestureState(Enum):
    """Simple gesture states."""

    IDLE = "idle"
    SCROLLING = "scrolling"


@dataclass
class VelocityUpdate:
    """Velocity update from controller."""

    velocity: Vec2D
    state: GestureState
    is_active: bool


class VelocityController:
    """Controls scroll gestures based on velocity.

    Much simpler than WheelController - just tracks touch state
    and passes velocity through. macOS handles all the complexity.
    """

    def __init__(self, min_velocity: float = 0.001):
        """Initialize controller.

        Args:
            min_velocity: Minimum velocity magnitude to start scrolling
        """
        self.min_velocity = min_velocity
        self.state = GestureState.IDLE
        self.was_touching = False

    def update(
        self, velocity: Vec2D | None, is_touching: bool, is_high_five: bool, timestamp_ms: int
    ) -> VelocityUpdate:
        """Update controller state.

        Args:
            velocity: Current velocity from tracker
            is_touching: Whether fingers are touching
            is_high_five: High-five gesture stops scrolling
            timestamp_ms: Current timestamp

        Returns:
            Velocity update with state
        """
        # High-five always stops
        if is_high_five:
            self.state = GestureState.IDLE
            return VelocityUpdate(velocity=Vec2D(0, 0), state=self.state, is_active=False)

        # Handle state transitions
        if self.state == GestureState.IDLE:
            # Start scrolling when touching with velocity
            if is_touching and velocity and velocity.magnitude > self.min_velocity:
                self.state = GestureState.SCROLLING

        elif self.state == GestureState.SCROLLING:
            # Stop when fingers lift
            if not is_touching and self.was_touching:
                # Let the scroll action know to end the gesture
                # macOS will handle momentum
                self.state = GestureState.IDLE

        self.was_touching = is_touching

        # Return current state
        return VelocityUpdate(
            velocity=velocity or Vec2D(0, 0),
            state=self.state,
            is_active=(self.state == GestureState.SCROLLING and is_touching),
        )

File: apps/web-ui/src/gestures/velocity-controller.ts
/**
 * Velocity-based scroll controller.
 * Ported from glide/gestures/velocity_controller.py
 */

import type { Vec2D } from './velocity-tracker';
import { vec2dZero, vec2dMagnitude } from './velocity-tracker';

export enum GestureState {
  IDLE = 'idle',
  SCROLLING = 'scrolling',
}

export interface VelocityUpdate {
  velocity: Vec2D;
  state: GestureState;
  isActive: boolean;
}

export class VelocityController {
  private minVelocity: number;
  private scrollDelayMs: number;
  state: GestureState = GestureState.IDLE;
  private wasTouching = false;
  private touchStartMs: number | null = null;

  constructor(minVelocity = 0.001, scrollDelayMs = 200) {
    this.minVelocity = minVelocity;
    this.scrollDelayMs = scrollDelayMs;
  }

  update(
    velocity: Vec2D | null,
    isTouching: boolean,
    isHighFive: boolean,
    timestampMs: number,
  ): VelocityUpdate {
    if (isHighFive) {
      this.state = GestureState.IDLE;
      this.touchStartMs = null;
      return { velocity: vec2dZero(), state: this.state, isActive: false };
    }

    if (this.state === GestureState.IDLE) {
      if (isTouching && velocity && vec2dMagnitude(velocity) > this.minVelocity) {
        if (this.touchStartMs == null) {
          this.touchStartMs = timestampMs;
        }
        if (timestampMs - this.touchStartMs >= this.scrollDelayMs) {
          this.state = GestureState.SCROLLING;
        }
      } else if (!isTouching) {
        this.touchStartMs = null;
      }
    } else if (this.state === GestureState.SCROLLING) {
      if (!isTouching && this.wasTouching) {
        this.state = GestureState.IDLE;
        this.touchStartMs = null;
      }
    }

    this.wasTouching = isTouching;

    return {
      velocity: velocity ?? vec2dZero(),
      state: this.state,
      isActive: this.state === GestureState.SCROLLING && isTouching,
    };
  }
}
